### ISSUE TITLE
Performance Degradation: vector reduce on Cuda

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#680]
Found a serious performance degradation due to cc9e146dff175805e2204d4dff9617882470b8b3

This needs to be resolved before we can merge into master. Costs about 2x on SPMV for miniFE matrix. This is related to issue #637 

Putting in as bug to have a high priority item.

**Comment by crtrott** [comment]
Looks like this newly introduced Reducer abstraction in impl/Kokkos_Reducer.hpp is the culprit. That thing is way to heavy weight for something to be called at the vector reduction level. Using its join instead of += makes performance drop from 17.6GFlop/s to 11.6 Gflop/s the rest of the drop seems to come from the new cuda_shfl functions which gets it down to 9.5GFlop/s.

**Comment by crtrott** [comment]
Ok looks like the compiler can actually get rid of all the intermediate stuff as long as the reduceop has a non-volatile overload of the join operator. If I do that then the new cuda_shfl operations don't hurt and some of the other stuff also doesn't hurt. Essentially thinning out the rest of the logic helped get some performance back even with the volatile join in there, but without it it didn't hurt in this case. But the finding indicates that the compiler has to work significantly harder to get this done, and might miss optimization possibilities because of the whole stack involved.

**Comment by crtrott** [comment]
Looks like its also affecting the host slowdown from 0.75s to 1.64s for my miniFE test between master and develop. This includes the fix for the reducer to have the non-volatile join.

**Comment by crtrott** [comment]
The host thing is less clear. The device thing is fixed (I hope).
